### PR TITLE
Make counts behave sensibly

### LINKED ISSUE #522: Turning off ergodic trimming gives different results
```python
        if self.ergodic_cutoff >= 1:
            # step 2. restrict the counts to the maximal strongly ergodic
            # subgraph
            self.countsmat_, mapping2 = _strongly_connected_subgraph(
                self.lag_time * raw_counts, self.ergodic_cutoff, self.verbose)
            self.mapping_ = _dict_compose(mapping, mapping2)
        else:
            # no ergodic trimming.
            self.countsmat_ = raw_counts
            self.mapping_ = mapping
```

case 1: multiply by lag time
case 2: will not get multiplied by lag time

makes a difference when using `prior_counts`

### PR DESCRIPTION
**Pull request by mpharrigan**
w.r.t. lag time, sliding window, and ergodic trimming

supercedes #532 and fixes #522 

Changes the behavior of `ergodic_cutoff` when `lag_time > 1` and `sliding_window == True`, but changes default behavior to be the same

### CODE REVIEW

**Comment by mpharrigan** [comment]
Commandline was failing with new options for `ergodic_cutoff`: in addition to a float, it should understand "on" and "off".

Fixed with latest commit. Involved changing some of the cmdline.py code

File: msmbuilder/msm/msm.py [hunk 1]
```diff
@@ -191,11 +199,9 @@ def fit(self, sequences, y=None):
         return self
 
     def _fit_mle(self, counts):
-        if self.ergodic_cutoff < 1:
-            with warnings.catch_warnings(record=True):
```
Comment (mpharrigan) [inline msmbuilder/msm/msm.py]: This warning was never being issued, btw

**Comment by mpharrigan** [comment]
I don't know why it fails on 3.3. Seems unrelated

File: msmbuilder/msm/msm.py [hunk 2]
```diff
@@ -54,24 +54,25 @@ class MarkovStateModel(BaseEstimator, _MappingTransformMixin, _SampleMSMMixin):
...
-    ergodic_cutoff : int, default=1
+    ergodic_cutoff : float or {'on', 'off'}, default='on'
         Only the maximal strongly ergodic subgraph of the data is used to build
         an MSM. Ergodicity is determined by ensuring that each state is
         accessible from each other state via one or more paths involving edges
         with a number of observed directed counts greater than or equal to
-        ``ergodic_cutoff``. Not that by setting ``ergodic_cutoff`` to 0, this
-        trimming is effectively turned off.
+        ``ergodic_cutoff``. Not that by setting ``ergodic_cutoff`` to 0 or
+        'off', this trimming is turned off.
```
Comment (rmcgibbo) [inline msmbuilder/msm/msm.py]: what is the difference between `on` and a number?

Comment (mpharrigan) [inline msmbuilder/msm/msm.py]: Are you asking or saying I should put it in the docs?

**Comment by rmcgibbo** [comment]
I looked this over, and +1 to merge.

**Comment by rmcgibbo** [comment]
@kyleabeauchamp do you want to look over this? (i've messed up this feature before... more eyes is probably better)

**Comment by kyleabeauchamp** [comment]
Will do

File: msmbuilder/commands/fit.py [hunk 3]
```diff
@@ -60,6 +60,12 @@ class MarkovStateModelCommand(FitCommand):
     _concrete = True
     _group = 'MSM'
 
+    def _ergodic_cutoff_type(self, erg):
```
Comment (kyleabeauchamp) [inline msmbuilder/commands/fit.py]: Silly question, but can someone point me to where I can figure out how this member function gets used?  I presume it's described in the numpydocs?

**Comment by kyleabeauchamp** [comment]
Another minor question: does the following block of code always work?  Note that it seems that classes can be *either* a list or a numpy array, but it is eventually broadcasted with `==`.  It seems to work on my machine, but I'm mosly curious.  

```
    if contains_nan:
        classes = classes[~np.isnan(classes)]
    if contains_none:
        classes = [c for c in classes if c is not None]

    n_states = len(classes)

    mapping = dict(zip(classes, range(n_states)))
    mapping_is_identity = np.all(classes == np.arange(n_states))
```

**Comment by kyleabeauchamp** [comment]
One more question.  Is it just me or is line 452 redundant:

```
    counts = np.zeros((n_states, n_states), dtype=float)
```
I ask because it could possibly be doubling the memory usage of a dense N^2 array...

File: msmbuilder/msm/core.py [hunk 4]
```diff
@@ -473,7 +478,13 @@ def _transition_counts(sequences, lag_time=1, sliding_window=True):
         shape=(n_states, n_states))
     counts = counts + np.asarray(C.todense())
 
-    return counts / float(lag_time), mapping
+    # The following condition isn't strictly necessary: if sliding window
```
Comment (kyleabeauchamp) [inline msmbuilder/msm/core.py]: I find it hard to understand what this comment is saying.  Is the code unnecessary?  If so, can we delete it?

File: msmbuilder/msm/msm.py [hunk 5]
```diff
@@ -54,24 +54,26 @@ class MarkovStateModel(BaseEstimator, _MappingTransformMixin, _SampleMSMMixin):
...
         direct symmetrization of the expected number of counts.
-    ergodic_cutoff : int, default=1
+    ergodic_cutoff : float or {'on', 'off'}, default='on'
         Only the maximal strongly ergodic subgraph of the data is used to build
         an MSM. Ergodicity is determined by ensuring that each state is
         accessible from each other state via one or more paths involving edges
         with a number of observed directed counts greater than or equal to
-        ``ergodic_cutoff``. Not that by setting ``ergodic_cutoff`` to 0, this
-        trimming is effectively turned off.
+        ``ergodic_cutoff``. Not that by setting ``ergodic_cutoff`` to 0 or
```
Comment (kyleabeauchamp) [inline msmbuilder/msm/msm.py]: "Not that" -> Just delete and say "By setting"

File: msmbuilder/commands/fit.py (hunk 3 above)
Comment (mpharrigan) [inline msmbuilder/commands/fit.py]: This is actually implemented in our `cmdline.py

Comment (mpharrigan) [inline msmbuilder/commands/fit.py]: It's called when instantiating the class ("klass") which the command wraps

File: msmbuilder/msm/core.py (hunk 4 above)
Comment (mpharrigan) [inline msmbuilder/msm/core.py]: The "if" could be removed, but it's more clear *why* we're dividing by lag time if it's kept

File: msmbuilder/msm/msm.py (hunk 5 above)
Comment (mpharrigan) [inline msmbuilder/msm/msm.py]: Good catch

**Comment by kyleabeauchamp** [comment]
How about something like this instead:

```
    #  If sliding window is False, this function will be called recursively
    #  with lag_time = 1 and strided trajectories, which gives the
    #   correct number of counts.  If sliding window is True, [...]
    # strided trajectories and lag_time = 1. Dividing by 1 is nothing.
    counts /= float(lag_time)
```

**Comment by kyleabeauchamp** [comment]
E.g. deleting the if statement and beefing up the comment

**Comment by mpharrigan** [comment]
Ok, what do you think of latest commit

**Comment by kyleabeauchamp** [comment]
+1

**Pull request merged**